include/up-cpp/utils/CallbackConnection.h

```diff
@@ -200,14 +304,97 @@ struct [[nodiscard]] CalleeHandle {
 	///     * False if the connection has been broken (i.e. This handle has
 	///       been reset/moved, or all other references to the connection
 	///       have been discarded)
-	explicit operator bool() const;
+	explicit operator bool() const {
+		auto locked_connection = connection_.lock();
+		return locked_connection && (*locked_connection);
+	}
 
 private:
-	std::shared_ptr<Conn> connection_;
+	std::weak_ptr<Conn> connection_;
 	std::shared_ptr<Callback> callback_;
 	std::optional<Cleanup> cleanup_;
 };
 
+/// @brief Thrown if a default constructed or reset() CallerHandle is called.
+///
+/// This does not occur if the connection was broken from the Callee end. When
+/// this is thrown, it represents a programmatic error around the use of a
+/// CallerHandle that needs to be corrected.
+struct BadCallerAccess : public std::logic_error {
+	template <typename... Args>
+	BadCallerAccess(Args&&... args)
+	    : std::logic_error(std::forward<Args>(args)...) {}
+};
+
+template <typename RT, typename... Args>
+struct [[nodiscard]] CallerHandle {
+	using Conn = Connection<RT, Args...>;
+
+	/// @brief Default construction results in an non-connected handle
+	CallerHandle() = default;
+
+	/// @brief Creates a connected handle. Only usable by Connection
+	CallerHandle(std::shared_ptr<Conn> connection,
+	             typename Conn::PrivateConstructToken)
+	    : connection_(connection) {
+		if (!connection_) {
+			throw BadConnection(
+			    "Attempted to create a connected CallerHandle with bad "
+			    "connection pointer");
+		}
+	}
+
+	/// @brief Drops this instance's copy of the handle, severing the
+	///        connection if no other CallerHandles represent the same handle.
+	void reset() { connection_.reset(); }
+
+	/// @brief Check if the connection is still valid.
+	///
+	/// @returns
+	///     * True if the connection is valid (i.e. at least one other
+	///       reference to the connection still exists).
+	///     * False if the connection has been broken (i.e. This handle has
+	///       been reset/moved, or all other references to the connection
+	///       have been discarded)
+	explicit operator bool() const { return connection_ && (*connection_); }
+
+	/// @throws BadCallerAccess if this handle has been default constructed OR
+	///         reset() has left it without a valid conneciton pointer.
+	auto operator()(Args&&... args) {
```

Comment:
These uses of auto are non-compliant to Autosar.  Here's a reference: Rule A7-1-5

Reply:
Autosar is not, to my knowledge, a requirement of up-cpp

Reply:
Reviewing Autosar A7-1-5, this use of `auto` would be allowed with a trailing return type to aid in type deduction:

> The auto specifier shall not be used apart from following cases: ... (4) to declare a function template using trailing return type syntax.

Based on that, a compliant fix would be:
```suggestion
	auto operator()(Args&&... args) -> decltype(*connection_)(std::forward<Args>(args)...)) {
```

(yes, the examples provided for a valid case 4 in Autosar A7-1-5 do use `decltype()` like that)

Personally, I find using a trailing return type in this context redundant - the exact same type deduction will occur regardless since the code that generates the return is copied twice.

Incidentally, Autosar has been merged with MISRA for C++17 and up based on what I've read, so the referenced rules are likely out of date for projects using anything after C++14.

Reply:
I think where we are at is that we need to see if we can get our hands on MISRA C++ 2023 to check the latest guidance for C++17 and C++20. We will also be asking for static analysis to be run against this code base. For the moment, I'd like to defer these changes until we have the appropriate references. There are several P0 and P1 issues waiting on the completion of this module. We will make a clean-up pass when possible to address MISRA/AUTOSAR compliance.

Reply:
Tracking request in #154